12.0 imp commown devices add module picking wizard

9184Lucky opened the pull request.

Commit:
Add an outward picking wizard on project task for not tracked products

Commit:
Add an inward picking wizard on project task for not tracked products

Commit:
Added tests

Commit:
i18n + l10n and fix typo

Commit:
Refactored contract so picking origin can be different than contract name

Also added a function to get task name in the form "Task-ID"

Commit:
Added a check on picking when a task enters a "check_picking_assigned" stage

Commit:
Refactoring to use constrain instead of write

fcayre commented on commown_devices/tests/common.py:
```diff
@@ -100,6 +102,48 @@ def adjust_stock(
 
+    def adjust_stock_notracking(self, product, qty=1.0, date="2000-01-01"):
+
+        location = self.env["stock.location"].create(
```
C'est très bizarre de créer un emplacement dans ce helper. Il vaudrait mieux la créer avant et la passer en argument, sinon on ne peut pas ajuster deux fois de suite le stock.

fcayre commented on commown_devices/tests/test_project_task.py:
```diff
@@ -365,3 +374,54 @@ def test_contract_resiliation_with_devices(self):
+                ]
+            )
+            .location_id
+        )
```
copier-coller => faire un petit helper (une fonction qui prend un produit suffit) qui retourne  la localisation actuelle. Ça me fait penser qu'il pourrait y avoir plusieurs quants et donc un plantage ? Peut-être qu'il faut au moins filtrer avec child_of le stock interne ?

9184Lucky replied:
+1 pour le helper ! Par contre pas sur de la condition sur le stock interne vu que si on va aussi se servir du helper pour réccupérer la localisation une fois envoyé au client...

fcayre commented on commown_devices/models/wizard_project_task_picking.py:
```diff
@@ -262,9 +262,9 @@ class ProjectTaskNoTrackingOutwardPickingWizard(models.TransientModel):
 
     @api.multi
-    def create_picking(self):
+    def create_picking(self, do_transfer=False):
```
Il est facile ici d'éviter de modifier le code juste pour les tests : il suffit d'appeler do_new_transfer(picking, date) avec from odoo.addons.commown_devices.models.common import do_new_transfer

9184Lucky replied:
Ok la remarque est marqué "outdated" et effectivement le code est de nouveau "create_picking(self)" cependant l'argument est encore présent dans les fonctions send_devices, send_devices_tracking_none, recieve_devices et recieve_devices_tracking_none (qui ont vocation à fusionner avec mon nouveau dev), est ce que on veut garder le parametre do_transfert à ce niveau là où je le supprime aussi ?

fcayre commented on commown_devices/i18n/fr.po:
```diff
@@ -159,7 +159,7 @@ msgstr "Diagnostique"
 #: model:ir.model.fields,field_description:commown_devices.field_project_task_outward_picking_wizard__display_name
 msgid "Display Name"
-msgstr "Nom affiché"
+msgstr "Afficher Nom"
```
Attention ce changement ne doit pas être pris.

fcayre commented on commown_devices/i18n/fr.po:
```diff
@@ -432,4 +438,5 @@ msgstr "project.task.inward.picking.wizard"
 msgid "project.task.outward.picking.wizard"
-msgstr "project.task.outward.picking.wizard"
+msgid "project.task.outward.picking.wizard"
+
```
C'est un bug ça je pense.

9184Lucky replied:
Je pense aussi, ça n'a plus l'air d'actualité...

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -45,23 +45,28 @@ def _compute_quant_ids(self):
         at the previous date.
         """
         dest_location = self.partner_id.get_or_create_customer_location()
+        if not origin:
```
Ici je trouve qu'on a le cul entre deux chaises : 
- soit on met comme condition "if origin is not None" et on évite le piège d'un argument passé falsy alors que c'est une erreur (genre l'entier 0)
- soit on assume et on utilise ci-dessous directement dans l'appel à _create_picking : origin=origin or self.name

Ici je pense que la 2ème solution convient.

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -76,70 +81,104 @@ def send_device_tracking_none(self, product, date=None, do_transfer=False):
             ]
         )
+        if not origin:
+            origin = self.name
```
idem

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -76,70 +81,104 @@ def send_device_tracking_none(self, product, date=None, do_transfer=False):
         at the previous date.
         """
+        if not origin:
+            origin = self.name
```
idem

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -76,70 +81,104 @@ def send_device_tracking_none(self, product, date=None, do_transfer=False):
         """
 
+        if not origin:
+            origin = self.name
```
idem

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -76,70 +81,104 @@ def send_device_tracking_none(self, product, date=None, do_transfer=False):
         self.ensure_one()
+
+        if not origin:
+            origin = self.name
```
idem

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -76,70 +81,104 @@ def send_device_tracking_none(self, product, date=None, do_transfer=False):
     ):
         self.ensure_one()
+        if not origin:
+            origin = self.name
```
idem

fcayre commented on commown_devices/models/crm_lead.py:
```diff
@@ -6,10 +6,6 @@
-def _assigned(picking):
-    return picking.state == "assigned"
-
-
```
Je suis pas fan de ce changement perso  : une lambda est créée et interprêtée à la volée, en terme de sobriété c'est pas ouf.

fcayre commented on commown_devices/models/wizard_project_task_picking.py:
```diff
@@ -216,6 +218,7 @@ def create_picking(self):
         return self.task_id.contract_id.receive_device(
             self.lot_id,
             self.env.ref("commown_devices.stock_location_devices_to_check"),
+            origin=self.task_id.get_id_name(),
```
get_id_name > get_name_for_origin ?

fcayre commented on commown_devices/models/project_task.py:
```diff
@@ -6,6 +6,8 @@
     "commown_devices.resiliated_stage",
 ]
 
+CHECK_PICKING_ASSIGNED_STAGE_XML_IDS = "commown_devices.sup_picking_ongoing_stage"
```
Typo dans le nom > pas de S à la fin. Je pense qu'une variable est inutile : elle n'est utilisée qu'une fois.

fcayre commented on commown_devices/models/wizard_project_task_picking.py:
```diff
@@ -283,11 +286,11 @@ class ProjectTaskNoTrackingInwardPickingWizard(models.TransientModel):
             origin=self.task_id.get_id_name(),
             date=self.date,
-            do_transfer=do_transfer,
+            do_transfer=False,
```
Ah, tu es revenu en arrière. Il faut essayer d'utiliser commit --amend dans ce cas pour ne pas surcharger inutilement la tâche du relecteur.

fcayre commented on commown_devices/tests/test_project_task.py:
```diff
@@ -39,6 +50,28 @@ def setUp(self):
+                {"name": "t1"},
+                {"name": "t2"},
+            ]
+        )
```
ça doit tenir sur une seule ligne de code non ?

fcayre commented on commown_devices/tests/test_project_task.py:
```diff
@@ -39,6 +50,28 @@ def setUp(self):
+            ]
+        )
+        self._create_xml_id(t1, "diagnostic_stage")
+        self._create_xml_id(t2, "resiliated_stage")
```
Je ne comprends pas pourquoi tu recrées dans les tests des entités censées être déjà créées dans les données. Est-ce que ton fichier de données est bien dans le __manifest__.py ? Car ça devrait carrément planter si oui...

fcayre commented on commown_devices/tests/test_project_task.py:
```diff
@@ -425,3 +463,40 @@ def test_outward_inward_notracking(self):
             self.env.ref("commown_devices.stock_location_devices_to_check"),
         )
+
+    def test_change_task_check(self):
```
test_change_task_check > test_change_stage_check plutôt ?

9184Lucky replied:
Carrément !

fcayre commented on commown_devices/tests/test_project_task.py:
```diff
@@ -425,3 +463,40 @@ def test_outward_inward_notracking(self):
+            [
+                ("product_id", "=", lot.id),
+                ("quantity", ">", 0),
+                ("location_id.complete_name", "ilike", "WH/STOCK/"),
```
wahou, ça marche ça ? Il est stored le champs complete_name ? Cool. J'aurais utilisé child_of, c'est un poil mieux que de se baser sur le nom des emplacements (qui pourrait changer).

9184Lucky replied:
Alors oui ça marche mais +1 pour le child of c'est plus propre !

fcayre commented on commown_devices/tests/test_project_task.py:
```diff
@@ -58,20 +48,7 @@ def setUp(self):
-        )
-        self._create_xml_id(t1, "diagnostic_stage")
-        self._create_xml_id(t2, "resiliated_stage")
+        self.ongoing_stage = self.env.ref("commown_devices.sup_picking_ongoing_stage")
```
OK vu idem : utiliser --amend pour ne pas surcharger le relecteur pour rien.

Commit:
Add an attribute on project stage to enable picking control

The attribute is dedicated to check if a picking attached to the task's contract is in 'assigned' state when entering a stage with that attribute set to True.

Commit:
Rename project task method

fcayre commented:
PR rebasée et mergée à la main.
